File: web-dashboard/src/lib/pdfExtractor.ts
export interface ExtractedMetadata {
  title?: string;
  authors?: string;
  journal?: string;
  year?: string;
  doi?: string;
  source: 'crossref' | 'pdf-text' | 'none';
}

/**
 * Read raw bytes from a PDF file as a latin1 string.
 */
async function extractRawText(file: File): Promise<string> {
  const buffer = await file.arrayBuffer();
  const bytes = new Uint8Array(buffer);

  let text = '';
  for (let i = 0; i < bytes.length; i++) {
    text += String.fromCharCode(bytes[i]);
  }

  return text;
}

/**
 * Extract PDF metadata fields (/Title, /Author, /Subject, /Creator).
 * PDFs store these as: /Title (The Paper Title) or /Title <hex>
 */
function extractPDFMetadata(raw: string): { title?: string; authors?: string } {
  const result: { title?: string; authors?: string } = {};

  // Match /Title (literal string)
  const titleMatch = raw.match(/\/Title\s*\(([^)]+)\)/);
  if (titleMatch) {
    const t = decodePDFString(titleMatch[1]);
    // Filter out generic/useless titles
    if (t.length > 3 && !/^untitled/i.test(t) && !/^microsoft/i.test(t) && t !== 'null') {
      result.title = t;
    }
  }

  // Match /Title <hex string>
  if (!result.title) {
    const titleHexMatch = raw.match(/\/Title\s*<([0-9A-Fa-f]+)>/);
    if (titleHexMatch) {
      const t = decodeHexString(titleHexMatch[1]);
      if (t.length > 3 && !/^untitled/i.test(t)) {
        result.title = t;
      }
    }
  }

  // Match /Author (literal string)
  const authorMatch = raw.match(/\/Author\s*\(([^)]+)\)/);
  if (authorMatch) {
    const a = decodePDFString(authorMatch[1]);
    if (a.length > 1 && a !== 'null') {
      result.authors = a;
    }
  }

  // Match /Author <hex string>
  if (!result.authors) {
    const authorHexMatch = raw.match(/\/Author\s*<([0-9A-Fa-f]+)>/);
    if (authorHexMatch) {
      const a = decodeHexString(authorHexMatch[1]);
      if (a.length > 1) {
        result.authors = a;
      }
    }
  }

  return result;
}

/**
 * Decode PDF escaped strings: \n, \r, \t, octal escapes, etc.
 */
function decodePDFString(s: string): string {
  return s
    .replace(/\\n/g, '\n')
    .replace(/\\r/g, '\r')
    .replace(/\\t/g, '\t')
    .replace(/\\\(/g, '(')
    .replace(/\\\)/g, ')')
    .replace(/\\\\/g, '\\')
    .replace(/\\(\d{1,3})/g, (_, oct) => String.fromCharCode(parseInt(oct, 8)))
    .trim();
}

/**
 * Decode hex-encoded PDF string (UTF-16BE or ASCII).
 */
function decodeHexString(hex: string): string {
  // Check for UTF-16BE BOM (FEFF)
  if (hex.startsWith('FEFF') || hex.startsWith('feff')) {
    let result = '';
    for (let i = 4; i < hex.length; i += 4) {
      const code = parseInt(hex.substring(i, i + 4), 16);
      if (code > 0) result += String.fromCharCode(code);
    }
    return result.trim();
  }

  // Plain ASCII hex
  let result = '';
  for (let i = 0; i < hex.length; i += 2) {
    const code = parseInt(hex.substring(i, i + 2), 16);
    if (code > 31) result += String.fromCharCode(code);
  }
  return result.trim();
}

/**
 * Extract readable text strings from PDF content streams.
 * Looks for text between parentheses in BT...ET blocks and Tj/TJ operators.
 */
function extractContentStrings(raw: string): string[] {
  const strings: string[] = [];

  // Match all literal strings used with text operators: (text) Tj or [(text)] TJ
  const textPattern = /\(([^)]{3,})\)\s*(?:Tj|TJ|'|")/g;
  let match;
  while ((match = textPattern.exec(raw)) !== null) {
    const decoded = decodePDFString(match[1]);
    if (decoded.length > 3 && /[a-zA-Z]/.test(decoded)) {
      strings.push(decoded);
    }
  }

  // Also grab longer literal strings that might be title/author even without operators
  const literalPattern = /\(([^)]{10,200})\)/g;
  while ((match = literalPattern.exec(raw)) !== null) {
    const decoded = decodePDFString(match[1]);
    if (decoded.length > 10 && /[a-zA-Z]{3,}/.test(decoded) && !/^[\d\s.]+$/.test(decoded)) {
      if (!strings.includes(decoded)) {
        strings.push(decoded);
      }
    }
  }

  return strings;
}

/**
 * Heuristic: try to identify the title from extracted text strings.
 * The title is usually one of the first substantial strings, often the longest.
 */
function guessTitle(strings: string[]): string | undefined {
  // Filter to candidate title strings
  const candidates = strings
    .slice(0, 30) // Look in the first 30 strings only (title is near the top)
    .filter(s =>
      s.length > 10 &&
      s.length < 300 &&
      !/^(abstract|introduction|references|acknowledgment|copyright|doi|arxiv|http|www\.|vol\.|issue|page|received|accepted|published|©)/i.test(s) &&
      !/^\d+$/.test(s) &&
      /[A-Za-z]{4,}/.test(s) // Must have actual words
    );

  if (candidates.length === 0) return undefined;

  // The title is often the longest string among the first few candidates
  const topCandidates = candidates.slice(0, 5);
  return topCandidates.reduce((a, b) => a.length >= b.length ? a : b);
}

/**
 * Heuristic: try to identify authors from extracted text strings.
 * Authors usually appear right after the title, contain commas/semicolons/and.
 */
function guessAuthors(strings: string[], title?: string): string | undefined {
  const titleIndex = title ? strings.findIndex(s => s === title) : -1;
  const searchStart = titleIndex >= 0 ? titleIndex + 1 : 0;
  const searchStrings = strings.slice(searchStart, searchStart + 15);

  for (const s of searchStrings) {
    // Skip if it looks like the title or a section header
    if (s === title) continue;
    if (/^(abstract|introduction|keywords|doi|http|©)/i.test(s)) continue;

    // Authors often have patterns like "Last, First" or "First Last" with commas/and
    const hasNamePattern = /[A-Z][a-z]+/.test(s);
    const hasSeparator = /[,;]|(\band\b)/.test(s);
    const notTooLong = s.length < 500;
    const notAllCaps = s !== s.toUpperCase();
    const hasMultipleWords = s.split(/\s+/).length >= 2;

    if (hasNamePattern && hasSeparator && notTooLong && notAllCaps && hasMultipleWords) {
      // Extra validation: should have at least 2 capitalized words
      const capWords = s.match(/[A-Z][a-z]+/g);
      if (capWords && capWords.length >= 2) {
        return s;
      }
    }
  }

  return undefined;
}

// --- Identifier finders ---

function findDOI(text: string): string | null {
  const doiPatterns = [
    /doi\.org\/(10\.\d{4,9}\/[^\s,;)}\]>"']+)/i,
    /doi[:\s]+\s*(10\.\d{4,9}\/[^\s,;)}\]>"']+)/i,
    /\b(10\.\d{4,9}\/[^\s,;)}\]>"']+)/i,
  ];

  for (const pattern of doiPatterns) {
    const match = text.match(pattern);
    if (match) {
      return match[1].replace(/[.)}\]>]+$/, '');
    }
  }
  return null;
}

function findArxivId(text: string): string | null {
  const patterns = [
    /arXiv[:\s]*(\d{4}\.\d{4,5}(?:v\d+)?)/i,
    /arxiv\.org\/abs\/(\d{4}\.\d{4,5}(?:v\d+)?)/i,
    /arxiv\.org\/pdf\/(\d{4}\.\d{4,5}(?:v\d+)?)/i,
  ];

  for (const pattern of patterns) {
    const match = text.match(pattern);
    if (match) return match[1];
  }
  return null;
}

function findISBN(text: string): string | null {
  const match = text.match(/(?:ISBN[:\s-]*)?(\d{3}[\s-]?\d[\s-]?\d{3}[\s-]?\d{5}[\s-]?\d)/i);
  if (match) return match[1].replace(/[\s-]/g, '');

  const match10 = text.match(/(?:ISBN[:\s-]*)?(\d[\s-]?\d{3}[\s-]?\d{5}[\s-]?[\dXx])/i);
  if (match10) return match10[1].replace(/[\s-]/g, '');

  return null;
}

function findYear(text: string): string | undefined {
  const currentYear = new Date().getFullYear();
  const yearMatch = text.match(/\b(19[5-9]\d|20[0-2]\d)\b/);
  if (yearMatch && parseInt(yearMatch[1]) <= currentYear + 1) {
    return yearMatch[1];
  }
  return undefined;
}

// --- API lookups ---

async function fetchFromCrossRef(doi: string): Promise<ExtractedMetadata | null> {
  try {
    const res = await fetch(
      `https://api.crossref.org/works/${encodeURIComponent(doi)}`,
      {
        headers: {
          'User-Agent': 'ScholarSync/1.0 (mailto:support@scholarsync.app)',
        },
      }
    );
    if (!res.ok) return null;

    const json = await res.json();
    const work = json.message;

    const authors = work.author
      ?.map((a: any) => `${a.family}${a.given ? ', ' + a.given : ''}`)
      .join('; ');

    const year = work.published?.['date-parts']?.[0]?.[0]
      || work['published-print']?.['date-parts']?.[0]?.[0]
      || work['published-online']?.['date-parts']?.[0]?.[0];

    const journal = work['container-title']?.[0]
      || work['short-container-title']?.[0];

    const title = work.title?.[0];

    return {
      title: title || undefined,
      authors: authors || undefined,
      journal: journal || undefined,
      year: year?.toString() || undefined,
      doi,
      source: 'crossref',
    };
  } catch {
    return null;
  }
}

async function fetchFromOpenLibrary(isbn: string): Promise<ExtractedMetadata | null> {
  try {
    const res = await fetch(`https://openlibrary.org/isbn/${isbn}.json`);
    if (!res.ok) return null;

    const data = await res.json();

    return {
      title: data.title || undefined,
      year: data.publish_date ? data.publish_date.match(/\d{4}/)?.[0] : undefined,
      source: 'crossref',
    };
  } catch {
    return null;
  }
}

// --- Main pipeline ---

/**
 * Extract metadata from a PDF:
 * 1. Read raw bytes
 * 2. Try DOI → CrossRef (best quality)
 * 3. Try arXiv ID → CrossRef
 * 4. Try ISBN → Open Library
 * 5. Fall back to PDF /Title + /Author metadata fields
 * 6. Fall back to heuristic text extraction from content streams
 */
export async function extractMetadataFromPDF(file: File): Promise<ExtractedMetadata> {
  try {
    const raw = await extractRawText(file);

    // --- Try API lookups first (highest quality) ---

    const doi = findDOI(raw);
    if (doi) {
      const data = await fetchFromCrossRef(doi);
      if (data && data.title) return data;
    }

    const arxivId = findArxivId(raw);
    if (arxivId) {
      const arxivDoi = `10.48550/arXiv.${arxivId}`;
      const data = await fetchFromCrossRef(arxivDoi);
      if (data && data.title) return data;
    }

    const isbn = findISBN(raw);
    if (isbn) {
      const data = await fetchFromOpenLibrary(isbn);
      if (data && data.title) return data;
    }

    // --- Fall back to local PDF extraction ---

    // 1. Try PDF metadata fields (/Title, /Author)
    const metadata = extractPDFMetadata(raw);

    // 2. Try content stream text extraction
    const contentStrings = extractContentStrings(raw);
    const guessedTitle = guessTitle(contentStrings);
    const guessedAuthors = guessAuthors(contentStrings, guessedTitle);
    const year = findYear(raw);

    // Merge: prefer metadata fields, fall back to heuristic guesses
    const title = metadata.title || guessedTitle;
    const authors = metadata.authors || guessedAuthors;

    const hasAnything = title || authors || doi || arxivId;

    return {
      title,
      authors,
      year,
      doi: doi || (arxivId ? `10.48550/arXiv.${arxivId}` : undefined),
      source: hasAnything ? 'pdf-text' : 'none',
    };
  } catch (err) {
    console.error('PDF extraction failed:', err);
    return { source: 'none' };
  }
}

File: ios-app/ScholarSync/ScholarSync/Models/PDFExtractor.swift
import Foundation

/// Extracts metadata identifiers (DOI, arXiv ID, ISBN) from raw PDF data,
/// then looks up full metadata via CrossRef or Open Library.
class PDFExtractor {
    static let shared = PDFExtractor()

    struct ExtractedMetadata {
        var title: String?
        var authors: String?
        var journal: String?
        var year: Int?
        var doi: String?
        var abstract: String?
        var source: String // "crossref", "pdf", "none"
    }

    /// Main extraction pipeline for a PDF file's raw data.
    func extract(from data: Data) async -> ExtractedMetadata {
        let rawText = extractRawText(from: data)

        // 1. Try DOI → CrossRef
        if let doi = findDOI(in: rawText) {
            if let paper = try? await CrossrefService.shared.fetchPaper(doi: doi) {
                return ExtractedMetadata(
                    title: paper.title,
                    authors: paper.authors,
                    journal: paper.journal,
                    year: paper.year,
                    doi: paper.doi,
                    abstract: paper.abstract,
                    source: "crossref"
                )
            }
        }

        // 2. Try arXiv ID → CrossRef via DOI
        if let arxivId = findArxivId(in: rawText) {
            let arxivDoi = "10.48550/arXiv.\(arxivId)"
            if let paper = try? await CrossrefService.shared.fetchPaper(doi: arxivDoi) {
                return ExtractedMetadata(
                    title: paper.title,
                    authors: paper.authors,
                    journal: paper.journal,
                    year: paper.year,
                    doi: paper.doi ?? arxivDoi,
                    abstract: paper.abstract,
                    source: "crossref"
                )
            }
        }

        // 3. Try ISBN → Open Library
        if let isbn = findISBN(in: rawText) {
            if let paper = try? await CrossrefService.shared.fetchPaper(doi: isbn) {
                return ExtractedMetadata(
                    title: paper.title,
                    authors: paper.authors,
                    journal: paper.journal,
                    year: paper.year,
                    doi: nil,
                    abstract: nil,
                    source: "crossref"
                )
            }
        }

        // 4. Fall back to PDF metadata fields and text heuristics
        let pdfMeta = extractPDFMetadata(from: rawText)
        let contentStrings = extractContentStrings(from: rawText)
        let title = pdfMeta.title ?? guessTitle(from: contentStrings)
        let authors = pdfMeta.authors ?? guessAuthors(from: contentStrings, title: title)
        let year = findYear(in: rawText)

        let hasAnything = title != nil || authors != nil
        return ExtractedMetadata(
            title: title,
            authors: authors,
            journal: nil,
            year: year,
            doi: findDOI(in: rawText),
            abstract: nil,
            source: hasAnything ? "pdf" : "none"
        )
    }

    // MARK: - Raw Text

    private func extractRawText(from data: Data) -> String {
        // Decode as latin1 to preserve all byte values
        return String(data: data, encoding: .isoLatin1) ?? ""
    }

    // MARK: - Identifier Finders

    private func findDOI(in text: String) -> String? {
        let patterns = [
            "doi\\.org/(10\\.\\d{4,9}/[^\\s,;)}\\]>\"']+)",
            "doi[:\\s]+\\s*(10\\.\\d{4,9}/[^\\s,;)}\\]>\"']+)",
            "\\b(10\\.\\d{4,9}/[^\\s,;)}\\]>\"']+)"
        ]
        for pattern in patterns {
            if let match = text.range(of: pattern, options: .regularExpression, range: text.startIndex..<text.endIndex) {
                var result = String(text[match])
                // Extract just the DOI part (capture group)
                if let doiRange = result.range(of: "10\\.\\d{4,9}/[^\\s,;)}\\]>\"']+", options: .regularExpression) {
                    result = String(result[doiRange])
                }
                // Clean trailing punctuation
                while result.last == "." || result.last == ")" || result.last == "}" || result.last == "]" || result.last == ">" {
                    result.removeLast()
                }
                return result
            }
        }
        return nil
    }

    private func findArxivId(in text: String) -> String? {
        let patterns = [
            "arXiv[:\\s]*(\\d{4}\\.\\d{4,5}(?:v\\d+)?)",
            "arxiv\\.org/abs/(\\d{4}\\.\\d{4,5}(?:v\\d+)?)",
            "arxiv\\.org/pdf/(\\d{4}\\.\\d{4,5}(?:v\\d+)?)"
        ]
        for pattern in patterns {
            if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) {
                let nsText = text as NSString
                let range = NSRange(location: 0, length: min(nsText.length, 50000))
                if let match = regex.firstMatch(in: text, range: range), match.numberOfRanges > 1 {
                    return nsText.substring(with: match.range(at: 1))
                }
            }
        }
        return nil
    }

    private func findISBN(in text: String) -> String? {
        let pattern = "ISBN[:\\s-]*(\\d{3}[\\s-]?\\d[\\s-]?\\d{3}[\\s-]?\\d{5}[\\s-]?\\d)"
        if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) {
            let nsText = text as NSString
            let range = NSRange(location: 0, length: min(nsText.length, 50000))
            if let match = regex.firstMatch(in: text, range: range), match.numberOfRanges > 1 {
                let isbn = nsText.substring(with: match.range(at: 1))
                return isbn.replacingOccurrences(of: "[\\s-]", with: "", options: .regularExpression)
            }
        }
        return nil
    }

    private func findYear(in text: String) -> Int? {
        let pattern = "\\b(19[5-9]\\d|20[0-2]\\d)\\b"
        if let regex = try? NSRegularExpression(pattern: pattern),
           let match = regex.firstMatch(in: text, range: NSRange(location: 0, length: min(text.count, 50000))),
           match.numberOfRanges > 1 {
            let yearStr = (text as NSString).substring(with: match.range(at: 1))
            return Int(yearStr)
        }
        return nil
    }

    // MARK: - PDF Metadata Fields

    private func extractPDFMetadata(from raw: String) -> (title: String?, authors: String?) {
        var title: String?
        var authors: String?

        // /Title (literal string)
        if let match = raw.range(of: "/Title\\s*\\(([^)]+)\\)", options: .regularExpression) {
            let full = String(raw[match])
            if let open = full.firstIndex(of: "("), let close = full.lastIndex(of: ")") {
                let t = String(full[full.index(after: open)..<close]).trimmingCharacters(in: .whitespacesAndNewlines)
                if t.count > 3 && !t.lowercased().hasPrefix("untitled") && !t.lowercased().hasPrefix("microsoft") {
                    title = t
                }
            }
        }

        // /Author (literal string)
        if let match = raw.range(of: "/Author\\s*\\(([^)]+)\\)", options: .regularExpression) {
            let full = String(raw[match])
            if let open = full.firstIndex(of: "("), let close = full.lastIndex(of: ")") {
                let a = String(full[full.index(after: open)..<close]).trimmingCharacters(in: .whitespacesAndNewlines)
                if a.count > 1 && a != "null" {
                    authors = a
                }
            }
        }

        return (title, authors)
    }

    // MARK: - Content Stream Extraction

    private func extractContentStrings(from raw: String) -> [String] {
        var strings: [String] = []

        // Match literal strings: (text) Tj or (text) TJ
        let pattern = "\\(([^)]{3,})\\)\\s*(?:Tj|TJ|'|\")"
        if let regex = try? NSRegularExpression(pattern: pattern) {
            let nsRaw = raw as NSString
            let searchLen = min(nsRaw.length, 100000)
            let matches = regex.matches(in: raw, range: NSRange(location: 0, length: searchLen))
            for match in matches where match.numberOfRanges > 1 {
                let s = nsRaw.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines)
                if s.count > 3 && s.rangeOfCharacter(from: .letters) != nil {
                    strings.append(s)
                }
            }
        }

        // Also grab longer literal strings
        let longPattern = "\\(([^)]{10,200})\\)"
        if let regex = try? NSRegularExpression(pattern: longPattern) {
            let nsRaw = raw as NSString
            let searchLen = min(nsRaw.length, 100000)
            let matches = regex.matches(in: raw, range: NSRange(location: 0, length: searchLen))
            for match in matches where match.numberOfRanges > 1 {
                let s = nsRaw.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines)
                if s.count > 10 && s.rangeOfCharacter(from: .letters) != nil && !strings.contains(s) {
                    strings.append(s)
                }
            }
        }

        return strings
    }

    private func guessTitle(from strings: [String]) -> String? {
        let candidates = strings.prefix(30).filter { s in
            s.count > 10 && s.count < 300 &&
            !s.lowercased().hasPrefix("abstract") &&
            !s.lowercased().hasPrefix("introduction") &&
            !s.lowercased().hasPrefix("references") &&
            !s.lowercased().hasPrefix("http") &&
            !s.lowercased().hasPrefix("doi") &&
            s.rangeOfCharacter(from: .letters) != nil
        }
        return candidates.prefix(5).max(by: { $0.count < $1.count })
    }

    private func guessAuthors(from strings: [String], title: String?) -> String? {
        let titleIdx = title.flatMap { t in strings.firstIndex(of: t) } ?? -1
        let start = titleIdx >= 0 ? titleIdx + 1 : 0
        let searchStrings = Array(strings.dropFirst(start).prefix(15))

        for s in searchStrings {
            if s == title { continue }
            if s.lowercased().hasPrefix("abstract") || s.lowercased().hasPrefix("http") { continue }

            let hasComma = s.contains(",") || s.contains(";") || s.lowercased().contains(" and ")
            let hasUpperCase = s.range(of: "[A-Z][a-z]+", options: .regularExpression) != nil
            let words = s.split(separator: " ")

            if hasComma && hasUpperCase && words.count >= 2 && s.count < 500 {
                return s
            }
        }
        return nil
    }
}
